apiserver/facades/controller/charmrevisionupdater/updater_test.go

```diff
@@ -4,136 +4,126 @@
-	anAuthoriser := s.authoriser
-	anAuthoriser.Controller = false
-	endPoint, err := charmrevisionupdater.NewCharmRevisionUpdaterAPI(s.State, s.resources, anAuthoriser)
-	c.Assert(endPoint, gc.IsNil)
-	c.Assert(err, gc.ErrorMatches, "permission denied")
-}
+func (s *updaterSuite) TestCharmhubNoUpdate(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
 
-func (s *charmVersionSuite) TestUpdateRevisions(c *gc.C) {
-	s.AddMachine(c, "0", state.JobManageModel)
-	s.SetupScenario(c)
+	state := makeState(c, ctrl, nil)
 
-	curl := charm.MustParseURL("cs:quantal/mysql")
-	_, err := s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+	state.EXPECT().AllApplications().Return([]charmrevisionupdater.Application{
+		makeApplication(ctrl, "ch", "postgresql", "charm-2", "app-2", 42),
+	}, nil).AnyTimes()
 
-	curl = charm.MustParseURL("cs:quantal/wordpress")
-	_, err = s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+	state.EXPECT().AddCharmPlaceholder(charm.MustParseURL("ch:postgresql-42")).Return(nil)
 
-	result, err := s.charmrevisionupdater.UpdateLatestRevisions()
+	updater, err := charmrevisionupdater.NewCharmRevisionUpdaterAPIState(state, nil, newFakeCharmhubClient)
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(result.Error, gc.IsNil)
 
-	curl = charm.MustParseURL("cs:quantal/mysql")
-	pending, err := s.State.LatestPlaceholderCharm(curl)
+	result, err := updater.UpdateLatestRevisions()
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(pending.String(), gc.Equals, "cs:quantal/mysql-23")
+	c.Assert(result.Error, gc.IsNil)
+}
 
-	// Latest wordpress is already deployed, so no pending charm.
-	curl = charm.MustParseURL("cs:quantal/wordpress")
-	_, err = s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+func (s *updaterSuite) TestCharmNotInStore(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
 
-	// Varnish has an error when updating, so no pending charm.
-	curl = charm.MustParseURL("cs:quantal/varnish")
-	_, err = s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+	state := makeState(c, ctrl, nil)
 
-	// Update mysql version and run update again.
-	app, err := s.State.Application("mysql")
-	c.Assert(err, jc.ErrorIsNil)
-	ch := s.AddCharmWithRevision(c, "mysql", 23)
-	cfg := state.SetCharmConfig{
-		Charm:      ch,
-		ForceUnits: true,
-	}
-	err = app.SetCharm(cfg)
+	state.EXPECT().AllApplications().Return([]charmrevisionupdater.Application{
+		makeApplication(ctrl, "ch", "varnish", "charm-5", "app-1", 1),
+		makeApplication(ctrl, "cs", "varnish", "charm-6", "app-2", 2),
+	}, nil).AnyTimes()
+
+	updater, err := charmrevisionupdater.NewCharmRevisionUpdaterAPIState(state, newFakeCharmstoreClient, newFakeCharmhubClient)
 	c.Assert(err, jc.ErrorIsNil)
 
-	result, err = s.charmrevisionupdater.UpdateLatestRevisions()
+	result, err := updater.UpdateLatestRevisions()
```

Comment:
We need to check something other than result.Error is nil I would have thought?
How can we check that UpdateLatestRevisions() was called. I see we expect a call to AllApplications(). Is there something else we can do in addition?
eg we set up the fake charm store to have wordpress and mysql charms with revisions. I can't see where we use those revisions in any checks

Reply:
Per voice discussion, this is actually correct as is. The result type only includes an Error field, so that's all there is to check for that. However, we know the code paths are executed because of the gomock `state.EXPECT().AddCharmPlaceholder()` calls -- charms that are going to be updated are added as placeholders to state. If that's not called (or called too many times or with the wrong args), gomock will fail the test.

Reply:
As far as revisions go, that's the `-23` part of an assertion like the following:

```
state.EXPECT().AddCharmPlaceholder(charm.MustParseURL("cs:mysql-23")).Return(nil)
```